Sources/NIOEmbedded/AsyncEmbeddedEventLoop.swift

```diff
@@ -0,0 +1,357 @@
+/// has no proper eventing mechanism. Instead, reads and writes are fully controlled by the
+/// entity that instantiates the `AsyncEmbeddedEventLoop`. This property makes `AsyncEmbeddedEventLoop`
+/// of limited use for many application purposes, but highly valuable for testing and other
+/// kinds of mocking. Unlike `EmbeddedEventLoop`, `AsyncEmbeddedEventLoop` is fully thread-safe and
+/// safe to use from within a Swift concurrency context.
+///
+/// Unlike `EmbeddedEventLoop`, `AsyncEmbeddedEventLoop` does require that user tests appropriately
+/// enforce thread safety. Used carefully it is possible to safely operate the event loop without
+/// explicit synchronization, but it is recommended to use `establishContext` in any case where it's
+/// necessary to ensure that the event loop is not making progress.
+///
+/// Time is controllable on an `AsyncEmbeddedEventLoop`. It begins at `NIODeadline.uptimeNanoseconds(0)`
+/// and may be advanced by a fixed amount by using `advanceTime(by:)`, or advanced to a point in
+/// time with `advanceTime(to:)`.
+///
+/// If users wish to perform multiple tasks at once on an `AsyncEmbeddedEventLoop`, it is recommended that they
+/// use `establishContext` to perform the operations.
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+public final class AsyncEmbeddedEventLoop: EventLoop, @unchecked Sendable {
+    /// The current "time" for this event loop. This is an amount in nanoseconds.
+    /// As we need to access this from any thread, we store this as an atomic.
+    private let _now = NIOAtomic<UInt64>.makeAtomic(value: 0)
+    internal var now: NIODeadline {
+        return NIODeadline.uptimeNanoseconds(self._now.load())
+    }
+
+    /// This is used to derive an identifier for this loop.
+    private var thisLoopID: ObjectIdentifier {
+        return ObjectIdentifier(self)
+    }
+
+    /// A dispatch specific that we use to determine whether we are on the queue for this
+    /// "event loop".
+    private static let inQueueKey = DispatchSpecificKey<ObjectIdentifier>()
+
+    // Our scheduledTaskCounter needs to be an atomic because we're going to access it from
+    // arbitrary threads. This is required by the EventLoop protocol and cannot be avoided.
+    // Specifically, Scheduled<T> creation requires us to be able to define the cancellation
+    // operation, so the task ID has to be created early.
+    private let scheduledTaskCounter = NIOAtomic<UInt64>.makeAtomic(value: 0)
+    private var scheduledTasks = PriorityQueue<EmbeddedScheduledTask>()
+
+    /// Keep track of where promises are allocated to ensure we can identify their source if they leak.
+    private let _promiseCreationStore = PromiseCreationStore()
+
+    // The number of the next task to be created. We track the order so that when we execute tasks
+    // scheduled at the same time, we may do so in the order in which they were submitted for
+    // execution.
+    private var taskNumber = UInt64(0)
```

Comment:
Should this only be accessed from `queue`?

Reply:
Yes, updated.